processing/src/main/java/io/druid/query/filter/LikeDimFilter.java

```
@@ -75,94 +79,113 @@ public LikeDimFilter(
+        return new LikeMatcher(prefix.toString(), SuffixStyle.MATCH_ANY, null, Pattern.compile(regex.toString()));
       } else {
-        return new LikeMatcher(prefix.toString(), SuffixStyle.MATCH_PATTERN, Pattern.compile(regexString));
+        return new LikeMatcher(
+            prefix.toString(),
+            SuffixStyle.MATCH_PATTERN,
+            Pattern.compile(suffixRegexString),
+            Pattern.compile(regex.toString())
+        );
       }
     }
 
-    public boolean matches(@Nullable final String s)
+    private static void addPatternCharacter(final StringBuilder patternBuilder, final char c)
     {
-      if (s == null) {
-        return prefix == null && (suffixStyle == SuffixStyle.MATCH_ANY
-                                  || suffixStyle == SuffixStyle.MATCH_EMPTY
-                                  || suffixPattern.matcher("").matches());
-      }
-
-      final String suffix;
-
-      if (prefix == null) {
-        suffix = s;
-      } else if (s.startsWith(prefix)) {
-        suffix = s.substring(prefix.length());
+      if (DEFINITELY_FINE.matcher(String.valueOf(c)).matches()) {
+        patternBuilder.append(c);
       } else {
-        return false;
+        patternBuilder.append("\\u").append(BaseEncoding.base16().encode(Chars.toByteArray(c)));
       }
+    }
+
+    public boolean matches(@Nullable final String s)
+    {
+      return pattern.matcher(Strings.nullToEmpty(s)).matches();
+    }
 
+    /**
+     * Checks if the suffix of "s" matches the suffix of this matcher. The first prefix.length characters
+     * of s are ignored. This method is useful if you've already independently verified the prefix.
+     */
+    public boolean matchesSuffixOnly(@Nullable final String s)
+    {
       if (suffixStyle == SuffixStyle.MATCH_ANY) {
         return true;
       } else if (suffixStyle == SuffixStyle.MATCH_EMPTY) {
-        return suffix.isEmpty();
+        return s == null || s.length() == prefix.length();
```

@gianm seems like if caller is feeding `null` here while prefix is not empty/`null`, it's a mistake in caller code, maybe should better throw something like NPE/IAE, rather than silently return true?